usaspending_api/etl/award_helpers.py

```diff
@@ -0,0 +1,116 @@
+from django.db import connection
+
+
+def update_awards(award_tuple=None):
+    """
+    Awards can have one or more transactions. We maintain some information
+    on the award model that needs to be updated as its child transactions
+    change. For example, an award's total obligated amount represents the
+    summary of its transaction's obligated amounts. Another example is a
+    series of fields (award type, awarding agency, etc.) that will always
+    be set to the value of the Award's most recent transaction.
+
+    This function keeps those awards fields synced with child transactions.
+    Obviously the raw SQL is not ideal. That said, the complex update
+    of award fields based on the earliest, latest, and aggregate values
+    of the child transactions was problematic to do in a set-based way
+    via the ORM. These updates do need to be set-based, as looping through
+    and updating individual award records would be an ETL bottleneck.
+    """
+
+    # common table expression for each award's latest transaction
+    sql_txn_latest = (
+        'txn_latest AS ('
+        'SELECT DISTINCT ON (award_id) * '
+        'FROM transaction ')
+    if award_tuple is not None:
+        sql_txn_latest += 'WHERE award_id IN %s '
+    sql_txn_latest += 'ORDER BY award_id, action_date DESC) '
+
+    # common table expression for each award's earliest transaction
+    sql_txn_earliest = (
+        'txn_earliest AS ('
+        'SELECT DISTINCT ON (award_id) * '
+        'FROM transaction ')
+    if award_tuple is not None:
+        sql_txn_earliest += 'WHERE award_id IN %s '
+    sql_txn_earliest += 'ORDER BY award_id, action_date) '
+
+    # common table expression for each award's summarized data
+    # (currently the only we summarize is federal_actio_obligation,
+    # but we can add more as necessar)
+    sql_txn_totals = (
+        'txn_totals AS ('
+        'SELECT award_id, SUM(federal_action_obligation) AS total_obligation '
+        'FROM transaction ')
+    if award_tuple is not None:
+        sql_txn_totals += 'WHERE award_id IN %s '
+    sql_txn_totals += 'GROUP BY award_id) '
+
+    # construct a sql query that uses the common table expressions
+    # defined above and joins each of them to their corresopnding
+    # award. the joined data from earliest, latest, and summarized
+    # transactions are used to update awards fields as appropriate
+    sql_update = 'WITH {}, {}, {}'.format(sql_txn_latest, sql_txn_earliest, sql_txn_totals)
+    sql_update += (
+        'UPDATE awards a '
+        'SET awarding_agency_id = l.awarding_agency_id, '
+        'certified_date = l.certified_date, '
+        'data_source = l.data_source, '
+        'date_signed = e.action_date, '
+        'description = l.description, '
```

Comment:
Geoff has informed me that description should come from the earliest transaction; but he said everything else looked fine :)